### DISCUSSION TITLE
Rendering groups the same way an activeSelection with preserveObjectStacking does?

### DISCUSSION THREAD

**Discussion by mattguest** [fabricjs/fabric.js#9092] [Q&A] [answered]
Hello all!

I am working on a tabletop map editor using FabricJS 5.3.1, and I love everything it brings to the table.
I have been stumped on something that I keep putting aside and coming back to, and I've reached my wits end.

I have some custom sorting going on when an object is moved, changing the draw order of objects on the canvas. I am resorting the canvas _objects array, which works great and allows me to sort objects in complex ways, faking layers and doing a video game style y-sort that puts objects towards the top of the canvas behind objects further down when you drag them around.

My problem is with groups. Groups are amazing for what I'm building, but for various reasons I need the individual objects in a group to render inline with all the other objects on the canvas so their draw order is correct.

I'll try to draw something here that explains this:
A - top object
B 
C
D - bottom object

If I group object A and C together and move the group around A needs to always be below B and D. C should draw on top of B, but below D. I hope that makes sense.

My first thought was "well, groups just won't work for this because they become one object", but the thing that's driving me nuts is that an activeSelection with preserveObjectStacking enabled on the canvas works exactly how I want this to work.

If I don't group the objects and just select A and C at the same time A will always be below B and C, and C will draw in-between B and D - exactly how I want grouping to work for my project.

Every search I do on the subject, and the Fabric code itself, shows that an ActiveSelection is just an extension of Group. I don't see anything specific in the ActiveSelection code to make it render objects differently than a Group does.

I keep coming close to writing my own grouping, but I'd be duplicating a ton of functionality and, well, active selections already do _exactly_ what I'm looking for.

I'd be grateful for any ideas you folks might have. Thanks!

**Comment by ShaMan123** [comment, marked as answer]
> If I group object A and C together and move the group around A needs to always be below B and D. C should draw on top of B, but below D. I hope that makes sense.

You mean you want the order to stay as you stated intially, right?

ActiveSelection is like Group apart from what you describe.
It is not in charge of rendering its children, canvas is.
And is what you want.

I strongly suggest you upgrade to v6 because Group has been completely rewritten and is ever more powerful, impacting ActiveSelection ofcourse #7670 

Regardless I will try to help but it will be on v6, you will need to understand how to port it if you are stuck on v5.
Look into:
- `_chooseObjectsToRender`
- `_renderObjects`

If you need to support nested selection with this kind of stacking that is more tricky but feasible in v6

**Reply by ShaMan123** [reply]
Note that I think there are difference between the versions in subtle parts of this logic
For example, in v5 objects in active selection still belong to canvas and exist in `_objects`, in v6 at the moment that is still the case but I am not sure I will keep it that way
Groups however extract their objects from canvas in v6 - read #7670

**Reply by ShaMan123** [reply]
Rule of thumb, using visual tricks is a good way to solve things with canvas.
Meaning you might have a logical state (like the object tree) that is different from the visual outcome.

**Reply by mattguest** [reply]
Thank you ShaMan123, this gives me a some things to look into and think on.

It looks like you understood what I am trying to do, but I suppose another way to have put what I'm trying to do is to allow logical groups (selecting and transforming one object selects and transforms the whole group it belongs to) while allowing the draw order of each individual object to be handled at the the canvas level like activeSelection does currently.

I'm going to work on understanding the functions you pointed out and think about your last statement of separating logical state from the visual outcome. One thought I had was to add the concept of 'saved selections' to my map editor. I can see a use for groups as they work now in my project as well. For instance combining different objects together to add a sword and a helmet to a character, or combining different rock pieces together to make a new mountain object. In those cases the user would expect the new object to be treated as a single entity as if it was drawn that way from the start.

Saved selections would be for things like prebuilt rooms, where you have a floor tile that should always be drawn underneath other objects in the scene, and furniture that should always be sorted among the other furniture objects in the whole scene. In that case I want all of the objects to still belong to the canvas so they can be sorted and drawn in the appropriate order. I don't know if a single dev's special use case should sway your decision on keeping the activeSelection objects in the canvas `_objects` array, but being able to select and move multiple objects without changing their draw order in the base canvas has been key to making fabric a viable option for my map editor.

Thanks again, I really appreciate you taking time out to answer!

If anyone stumbles across this and is curious, here's a demo video of the map editor I'm using FabricJS for:
https://www.youtube.com/watch?v=3WoLt3Lbipo&list=TLGGrt1OqvF1n-QxNTA3MjAyMw&t=213s

I appreciate all of the hard work that goes into building something like fabricjs. I have been able to achieve something really special in a relatively short time of nights and weekends and look forward to showcasing it here when it's ready.

**Reply by ShaMan123** [reply]
Very cool
I have been wotkinh too long on collaboration solutions which are visually boring
Your app likes good
Don't worry,if I remove it it means I have found a much better and flexible solution for the greater good
Fabric needs to become accessible for all
